### DISCUSSION TITLE
COBREXA 2.0

### DISCUSSION THREAD

**Discussion by stelmo** [LCSB-BioCore/COBREXA.jl#671] [General] [answered]
Let's add thoughts and ideas for COBREXA 2.0 here:

### Breaking changes (v2.0)
- [x] Currently we use `reactions` and `fluxes` interchangeably and actually mean `variables`. We should not do this, as future models will certainly predict more than just fluxes, e.g. enzyme constrained models. See #613 as well as #633 
- [x] Add a result type wrapper to allow analysis functions pipes #763 
- [x] Make function discoverability easier by reorganizing everything into modules #663 
- [x] Consistently name kwargs #623 
- [x] Rename `MetabolicModel` to `AbstractMetabolicModel`, this is better Julia style. See #687 
- [x] Rename analysis functions: GeckoModel and SMomentModel should just be EnzymeConstrainedModels
- [x] Change the names of CoreModel, CoreCoupledModel, and StandardModel
- [ ] ~Combine `CoreModel` and `CoreCoupledModel`~ MatrixModel should get 🪓ed
- [x] Rename all `grr` to `gra`
- [x] Deprecate and remove moment modifications. See #686
- [x] Remove the `@add_reaction` macro - just use the `add_reaction` functions like a human #697
- [x] Rename `reaction_gene_association` to `reaction_gene_associations` #721 
- [ ] Improve the model construction functions. Make it easier to construct a _sensible_ model. In particular, using non-human readable IDs will basically ruin lots of our existing `looks_like` functions and tutorials etc.  #652 starts along this path
- [x] Fix and make community model construction more robust #591 
- [x] Make convenience functions for enzyme constrained models with ObjectModel #724 
- [x] Remove sparse matrix from `show` and pretty printing.. it usually doesn't display nicely and just clutters everything #698 
- [x] Fix references to `GeneAssociations` when they should be `GeneAssociationsDNF` e.g. in `MatrixModel` docstrings
- [ ] Add `with_` functions or wrappers for model extensions. E.g. a smooth way to add coupling constraints, that could be nested etc.
- [x] Pick between `reverse` and `backward` keywords (currently mixed usage) #722 
- [x] After module re-org, no functions are automatically exported. E.g. to use `load_model` I first need to do `using COBREXA.IO`. Maybe we should export everything by default to make the user experience "faster"? #716 
- [ ] Improve user experience with sampling #657 #665 
- [x] Convert MMDF to a wrapper #614 
- [x] Remove ability to construct reactions like this: `1.0 * nadh + 4.0 * h_c + q8 ← 1.0 * q8h2 + 1.0 * nad + 3.0 * h_p` #718 
- [x] Add `ReactionForward`, `ReactionBackward`, `ReactionBidirectional` as overloads to `Reaction` instead of the current symbol passing approach #740
- [x] Make pFBA is pipe-able #747
- [ ] Check that other analysis functions are pipe-able
- [x] Add fields listed in Carey 2020 to ObjectModel to make the structure conform to community standards #750
- [x] Ensure all the `with_` functions have analogs for all all ObjectModel reconstruction functions. E.g. currently `with_added_reactions` does not exist for ObjectModel #739 
- [x] `change_constraint` has kwarg `lower_bound` instead of `lower` - pick one for these types of functions #746
- [x] name constraints in gecko to make modification of opt problem easier
- [x] Deprecate crowding (they won't work with a general metabolic model) #725
- [ ] Allow loopless to be applied to any model e.g. enzyme constrained models #745 
- [x] Parse gene reaction rules more robustly #711 
- [ ] Extend summary functions with semantics
- [ ] Make GRR DNF access optional for models that do not require it, for example SBML can evaluate KOs directly on the boolean formula. (IMPORTANT.) 
- [ ] Convert `knockout` to a wrapper to actually work with reactions instead of plain variables.
- [ ] ~Be able to import and run enzyme constrained analysis on [these matlab models](https://github.com/SysBioChalmers/ecModels) through a general purpose enzymewrappermodel (Chalmer matlab EC models).~ Not necessary, construct EC models sensibly
- [ ] Consistently use the same delimiter for IDs and make it a const: `const delim = "#"`
- [x] Rework community fba #764 
- [x] Add ability to add multiple capacity constraints to `SimplifiedEnzymeModel` #772
- [ ] ~Add traits somehow with https://github.com/oxinabox/Tricks.jl/~ With Julia 1.10 this would be in the base language, delay until then
- [ ] Delete ObjectModel 🪓 
- [ ] Delete MatrixModel 🪓 
- [ ] Add methods to build a model directly with SBMLModel
- [ ] Add support for ME models (read in SBTab files) #664

### Close longstanding issues
- [ ] Enzyme models with mangled IDs #633
- [ ] Model conversion issues #709
- [ ] Each function should have an example and test in the docstring #73

### Non-breaking changes (v.2.1+)
- [ ] Add units used by enzyme constrained functions to the docstrings to remove ambiguity (probably also for the community stuff)
- [ ] better error messages for model reading, indicate which reaction/gene/metabolite breaks the IO or convert function
- [x] Pretty printing for `ModelWithResult` type, see #790 
- [ ] ~Need a way to save ObjectModels that have been extended with e.g. kcats etc...~ ObjectModel needs to be 🪓ed
- [ ] Add precompile statements to take advantage of native code caching 
- [ ] ~Use ReadableRegex instead of normal regex~ Not worth the effort
- [ ] Implement smetana for community analysis see [here for the method](https://www.pnas.org/doi/abs/10.1073/pnas.1421834112)

**Comment by stelmo** [comment]
Here is a potential option we can consider for a dramatic change:
```julia
#=
Is this a better way to encapsulate the default cobrexa model types?

Note, I assume we have renamed our abstract type to AbstractMetabolicModel,
thus JSONModel, SBMLModel, MATModel <: AbstractMetabolicModel
=#

abstract type AbstractReaction end
abstract type AbstractMetabolite end
abstract type AbstractGene end

# this should be a subtype of AbstractMetabolicModel but I don't know how to code that in...
struct MetabolicModel{R, M, G} where {R<:AbstractReaction, M<:AbstractMetabolite, G<:AbstractGene}
    id::String
    reactions::Vector{R}
    metabolites::Vector{M}
    genes::Vector{G}
end

# Implement CoreModel like this:
const CoreModel = MetabolicModel{CoreReaction, CoreMetabolite, CoreGene}

struct CoreReaction <: AbstractReaction 
    id :: String 
    column :: SparseVec
    lb :: Float64
    ub :: Float64
    objective_coefficient :: Maybe{Float64}
    coupled :: SparseVec
end

struct CoreMetabolite <: AbstractMetabolite
    id :: String # matches order of column in CoreReaction 
end

struct CoreGene <: AbstractGene
    id :: String # matches order of 
end

# Implement StandardModel like this:
const StandardModel = MetabolicModel{GeneralizedReaction{???}, ...} # here ??? depends on what you want 

struct GeneralizedReaction{T} <: AbstractReaction 
    id :: String
    metadata :: Vector{Pair{Symbol, T}}
end

#= 
metadata1 = [
    Pair(:name => "Pyruvate formate lyase"),
    Pair(:metabolite => ("accoa_c", 1)),
    Pair(:metabolite => ("coa_c", -1)),

    Pair(:upper_bound => 1000.0),

    Pair(:isozyme => (b0902, b0903)),
    Pair(:isozyme => (b0902, b3114)),
]

r1 = GeneralizedReaction("rxn1", metadata1)
=#

#= 
metadata2 = [
    Pair(:name => "Pyruvate formate lyase"),
    Pair(:metabolite => ("accoa_c", 1)),
    Pair(:metabolite => ("coa_c", -1)),

    Pair(:lower_bound => 0.0),


    Pair(:isozyme => ((1, b0902), (2, b0903))),
    Pair(:isozyme => ((3, b0902), (2, b3114))),
]

r2 = GeneralizedReaction("rxn2", metadata2)
=#

#= 
Then r1 and r2 have different types depending on what your model needs. 
=#
```

**Comment by exaexa** [comment]
Just a reminder to self -- @mihal-sysbio pointed out we should care about naming our variant of Gecko somehow better (which is reasonable to avoid confusion). It might be GeckoLite (need to verify).

**Reply by stelmo** [reply]
Let's just called it the `EnzymeConstrainedModel`... It seems they want GECKO to mean the pipeline where they estimate kcats for the enzyme constrained models instead of the algorithm to solve the problem. Since we won't support Moment anymore there is no conflict with using such a broad term.

**Reply by stelmo** [reply]
This will be addressed by #719 and #720

**Reply by mihai-sysbio** [reply]
FYI GECKO has _full_ and _light_ forms already, and it's at [`v3`](https://github.com/SysBioChalmers/GECKO#about-gecko-3) now.

**Comment by exaexa** [comment]
CODE FREEZE FOR 2.0: Feb 10th

**Reply by stelmo** [reply]
Probably a good idea to push this another week or two

**Reply by exaexa** [reply]
ok pushed. Feb 17th. :D

**Reply by stelmo** [reply]
Push another week to 24th

**Reply by exaexa** [reply, marked as answer]
spoiler: we redone `next` completely.

**Comment by mihai-sysbio** [comment]
Another thing to consider adding: more documentation. It was rather difficult for me to understand there are different types of Models that what can and cannot be done with each so that I do right conversion.
Also if versions appear, perhaps it would be possible to keep old documentation available online, e.g., with a dropdown tag with the release version.
On a more philosophical note, should model types be related to file formats or to the type purpose? I'm thinking of JSONModel and SBMLModel which both can take in a genome-scale metabolic model, but also that SBML can handle many other model types.

**Reply by exaexa** [reply]
This is very true. We're vastly reworking the whole thing in `next` branch. ConstraintTrees should make the necessity of multiple different model structures basically obsolete. We'll preserve the filetype models, but only as information sources/sinks.

For the documentation, there's a decision to have 100% coverage ONLY by stuff being rendered in literate notebooks, so hopefully that will help. We already did almost the same (except for a few corner cases) for CTs here: https://cobrexa.github.io/ConstraintTrees.jl/stable/

**Reply by exaexa** [reply]
PS the "old documentation" should be present, the dropdown should be in the lower left corner in the docs.
